File: prog/1stPartyLibs/quirrel/quirrel/bench/js/tree.js
// primes test

function Node(x) {
    this.x = x
    this.y = Math.random()
    this.left = null
    this.right = null
}


function merge(lower, greater) {
    if (lower === null) {
    	return greater
    }

    if (greater === null) {
    	return lower
    }

    if (lower.y < greater.y) {
        lower.right = merge(lower.right, greater)
        return lower
    } else {
        greater.left = merge(lower, greater.left)
        return greater
    }
}

function splitBinary(orig, value) {
    if (orig === null) {
    	return [null, null]
    }

    if (orig.x < value) {
        const splitPair = splitBinary(orig.right, value)
        orig.right = splitPair[0]
        return [orig, splitPair[1]]
    } else {
        const splitPair = splitBinary(orig.left, value)
        orig.left = splitPair[1]
        return [splitPair[0], orig]
    }
}

function merge3(lower, equal, greater) {
    return merge(merge(lower, equal), greater)
}


function SplitResult(lower, equal, greater) {
    this.lower = lower
    this.equal = equal
    this.greater = greater
}


function split(orig, value) {
    var lower_eqg = splitBinary(orig, value)
    var eq_greater = splitBinary(lower_eqg[1], value + 1)
    return new SplitResult(lower_eqg[0], eq_greater[0], eq_greater[1])
}


function Tree()
{
    this.root = null
}

Tree.prototype.has_value = function(x) {
    var splited = split(this.root, x)
    var res = splited.equal !== null
    this.root = merge3(splited.lower, splited.equal, splited.greater)
    return res
}

Tree.prototype.insert = function(x) {
    var splited = split(this.root, x)
    if (splited.equal === null) {
        splited.equal = new Node(x)
    }
    this.root = merge3(splited.lower, splited.equal, splited.greater)
}

Tree.prototype.erase = function(x) {
    var splited = split(this.root, x)
    this.root = merge(splited.lower, splited.greater)
}

function testTree() {
    var tree = new Tree()
    var cur = 5;
    var res = 0;

    for (var i = 1; i < 1000000; ++i) {
        var a = i % 3;
        cur = (cur * 57 + 43) % 10007;
        if (a === 0) {
            tree.insert(cur);
        } else if (a === 1) {
            tree.erase(cur);
        } else if (a == 2) {
            res += tree.has_value(cur) ? 1 : 0;
        }
    }
	return res;
}

// infrastructure

function timeStamp() {
    return Date.now();
}


function profile(tname,cnt,testFn) {
	var t = 100500;
	var count = cnt
	while ( count>0 ) {
		var t0 = timeStamp();
		testFn();
		var t1 = timeStamp();
		var dt = t1 - t0;
		t = Math.min(dt, t);
		count --;
	}
	t /= 1000.0;
	var msg = '"'+tname+'", '+t+', '+cnt;
    print(msg)
}

function performance_tests() {
	profile("tree",1,function(){
		testTree();
	});
	timeStamp();
}

performance_tests();

File: prog/1stPartyLibs/quirrel/quirrel/bench/lua/tree.lua
--Copyright and license: https://github.com/frol/completely-unscientific-benchmarks

local function merge(lower, greater)
    if not lower then
        return greater
    end

    if not greater then
        return lower
    end

    if lower.y < greater.y then
        lower.right = merge(lower.right, greater)
        return lower
    else
        greater.left = merge(lower, greater.left)
        return greater
    end
end

local function split_binary(orig, value)
    if not orig then
        return nil, nil
    end

    if orig.x < value then
        local origRight, origLeft = split_binary(orig.right, value)
        orig.right = origRight
        return orig, origLeft
    else
        local origRight, origLeft = split_binary(orig.left, value)
        orig.left = origLeft
        return origRight, orig
    end
end

local function merge3(lower, equal, greater)
    return merge(merge(lower, equal), greater)
end

local function split(orig, value)
    local lower, equalGreater = split_binary(orig, value)
    local equal, greater = split_binary(equalGreater, value + 1)
    return lower, equal, greater
end

local function tree_has_value(tree, x)
    local lower, equal, greater = split(tree.root, x)
    local res = equal ~= nil
    tree.root = merge3(lower, equal, greater)
    return res
end

local function tree_insert(tree, x)
    local lower, equal, greater = split(tree.root, x)
    if not equal then
        equal = {
            x = x,
            y = math.random(0, 2^31),
            left = nil,
            right = nil
        }
    end
    tree.root = merge3(lower, equal, greater)
end

local function tree_erase(tree, x)
    local lower, equal, greater = split(tree.root, x)
    tree.root = merge(lower, greater)
end

loadfile("profile.lua")()

local function main()
    local tree = {
        root = nil
    }
    local cur = 5
    local res = 0

    for i = 1, 1000000 do
        local a = i % 3
        cur = (cur * 57 + 43) % 10007
        if a == 0 then
            tree_insert(tree, cur)
        elseif a == 1 then
            tree_erase(tree, cur)
        elseif a == 2 then
            if tree_has_value(tree, cur) then
                res = res + 1
            end
        end
    end
    --print(res)
end

io.write(string.format("\"tree\", %.8f, 10\n", profile_it(10, function () main() end)))
